### PR TITLE
feat(app, app-shell): desktop: software update flow rework

### PR DESCRIPTION
**Pull request by mjhuff**
Partially closes RAUT-501

# Overview

This PR reworks the desktop app's software update flow. It includes a new download modal, and a toast to alert when a software update is available & successfully installed upon relaunch. It also adds a new download percentage shell-to-redux pipe. Additionally, Config typing is restructured to prevent infinite type depth errors. 

This PR does NOT implement a v19 config file with the hasJustUpdated flag defaulted to true. This is to prevent first time users from installing the product and then seeing a "successfully updated" toast. However, this does mean that users migrating to v19 config will not see the success toast until their subsequent update after migrating to Configv19. 

Here is a test build (test on an M1 mac):
https://builds.opentrons.com/app/Opentrons-v7.0.0-alpha.4-mac-b33793-app_desktop-foolproof-software-update-flows-app-build.dmg

### Future Improvements

There are interesting download/install edge cases that could result in incorrect UX feedback, however, these situations are likely uncommon enough to warrant addressing in the future.

### Update

https://github.com/Opentrons/opentrons/assets/64858653/e04dd15a-0127-4914-8996-7b930a3e8677

### Error 

https://github.com/Opentrons/opentrons/assets/64858653/fb064c62-b1cc-412e-bfb7-a24f81cbd859

### Remind Me Later

https://github.com/Opentrons/opentrons/assets/64858653/28f7e236-a212-4282-bc32-ac9f0b08d66c

# Test Plan

NOTE: Testing can be finicky if the newest version is already downloaded. Error messages within the rendered error modal should help create proper testing conditions.

- Verified all features (minus the actual update) work.
- Verified update works on M1 macs (thank you to all who helped!)
- Independently verified app success toast displays after an update (hard to verify in current flows, since upgrading app will remove the success toast). 

# Changelog

- Overhauled update flows/modal. New progress bar included.
- Added a download percentage shell-to-redux pipe. 
- Added toast support for updating reminders/success.
- Created Configv19 that now includes the hasJustUpdated flag.
- Restructured config typing to prevent infinite depth type related errors.

# Review requests
Test on an M1 Mac. There are potential build issues for users without M1 Macs (not relating to this PR).

# Risk assessment
medium

### CODE REVIEW

File: app-shell-odd/src/config/__tests__/migrate.test.ts [hunk 1]
```diff
@@ -7,63 +7,74 @@ import {
   MOCK_CONFIG_V16,
   MOCK_CONFIG_V17,
   MOCK_CONFIG_V18,
+  MOCK_CONFIG_V19,
 } from '../__fixtures__'
 import { migrate } from '../migrate'
 
+const NEWEST_VERSION = 19
```
Comment (jerader) [inline app-shell-odd/src/config/__tests__/migrate.test.ts:14]: this is good idea lol 🤩

File: app-shell/src/__tests__/update.test.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@
 import * as ElectronUpdater from 'electron-updater'
 import { registerUpdate } from '../update'
 import * as Cfg from '../config'
+import { UPDATE_VALUE } from '@opentrons/app/src/redux/config'
```
Comment (jerader) [inline app-shell/src/__tests__/update.test.ts]: import can go a few lines higher

File: app/src/organisms/Alerts/__tests__/Alerts.test.tsx [hunk 3]
```diff
@@ -6,6 +6,7 @@ import { Alerts } from '..'
 import { AnalyticsSettingsModal } from '../../AnalyticsSettingsModal'
 import { U2EDriverOutdatedAlert } from '../U2EDriverOutdatedAlert'
 import { UpdateAppModal } from '../../UpdateAppModal'
+import { TOAST_ANIMATION_DURATION } from '../../../atoms/Toast'
```
Comment (jerader) [inline app/src/organisms/Alerts/__tests__/Alerts.test.tsx]: import can go higher

File: app/src/organisms/Alerts/index.tsx [hunk 4]
```diff
@@ -1,39 +1,82 @@
...
+import { useTranslation } from 'react-i18next'
 import head from 'lodash/head'
 
 import * as AppAlerts from '../../redux/alerts'
+import { getHasJustUpdated, toggleConfigValue } from '../../redux/config'
 import { AnalyticsSettingsModal } from '../AnalyticsSettingsModal'
 import { UpdateAppModal } from '../UpdateAppModal'
 import { U2EDriverOutdatedAlert } from './U2EDriverOutdatedAlert'
+import { useToaster } from '../ToasterOven'
+import { SUCCESS_TOAST, WARNING_TOAST } from '../../atoms/Toast'
```
Comment (jerader) [inline app/src/organisms/Alerts/index.tsx]: import can go higher

File: app/src/organisms/Alerts/index.tsx [hunk 5]
```diff
@@ -1,39 +1,82 @@
...
           own render; move its logic into `state.alerts` */}
       <AnalyticsSettingsModal />
 
-      {activeAlert === AppAlerts.ALERT_U2E_DRIVER_OUTDATED ? (
+      {activeAlertId === AppAlerts.ALERT_U2E_DRIVER_OUTDATED ? (
         <U2EDriverOutdatedAlert dismissAlert={dismissAlert} />
-      ) : activeAlert === AppAlerts.ALERT_APP_UPDATE_AVAILABLE ? (
-        <UpdateAppModal dismissAlert={dismissAlert} />
+      ) : showUpdateModal ? (
+        <UpdateAppModal closeModal={() => setShowUpdateModal(false)} />
```
Comment (jerader) [inline app/src/organisms/Alerts/index.tsx]: nit: not introduced in this pr but this nested ternary is a bit hard to read

File: app/src/organisms/UpdateAppModal/__tests__/UpdateAppModal.test.tsx [hunk 6]
```diff
@@ -1,48 +1,49 @@
 import * as React from 'react'
-import { Link as InternalLink } from 'react-router-dom'
-import { mountWithStore, BaseModal, Flex, Icon } from '@opentrons/components'
+import { fireEvent } from '@testing-library/react'
 import * as Shell from '../../../redux/shell'
-import { ErrorModal } from '../../../molecules/modals'
-import { ReleaseNotes } from '../../../molecules/ReleaseNotes'
-import { UpdateAppModal } from '..'
+import { i18n } from '../../../i18n'
+import { renderWithProviders } from '@opentrons/components'
```
Comment (jerader) [inline app/src/organisms/UpdateAppModal/__tests__/UpdateAppModal.test.tsx:5]: import can go higher

File: app/src/redux/config/schema-types.ts [hunk 7]
```diff
@@ -99,101 +99,100 @@ export interface ConfigV0 {
   devInternal?: FeatureFlags
 }
 
-export interface ConfigV1 extends Omit<ConfigV0, 'version' | 'discovery'> {
+export type ConfigV1 = Omit<ConfigV0, 'version' | 'discovery'> & {
```
Comment (jerader) [inline app/src/redux/config/schema-types.ts:99]: hmm, i always get confused when to use `type` vs `interface`, i also thought `lint` gives you an error if you incorrectly call something a `type` 🤔 , but if lint isn't mad then this is fine ⭐

**Review by jerader — COMMENTED** [review]
left some clean up comments. Also nit: since you touched `app-shell` directory, you can include it in your PR title: `feat(app, app-shell): desktop: software update flow rework`

File: app/src/redux/config/schema-types.ts (hunk 7 above)
Comment (mjhuff) [reply app/src/redux/config/schema-types.ts:99]: Yeah, it confuses me too - definitely some under the hood TS type checking that's beyond me. For some reason using type instead of interface + extends doesn't trigger the TS "infinitely deep" error that occurs past a certain point (it seems like it's ~19 "interface extends") 🤷

File: app-shell-odd/src/config/__tests__/migrate.test.ts (hunk 1 above)
Comment (koji) [reply app-shell-odd/src/config/__tests__/migrate.test.ts:14]: nice

File: app-shell-odd/src/discovery.ts [hunk 8]
```diff
@@ -42,7 +42,7 @@ interface DiscoveryStore {
   services?: LegacyService[]
 }
 
-let config: Config['discovery']
+let config: ConfigV1['discovery']
```
Comment (brenthagen) [inline app-shell-odd/src/discovery.ts:45]: why change from the latest `Config`?

File: app-shell/src/log.ts [hunk 9]
```diff
@@ -25,7 +25,7 @@ const FILE_OPTIONS = {
   tailable: true,
 }
 
-let config: Config['log']
+let config: ConfigV0['log']
```
Comment (brenthagen) [inline app-shell/src/log.ts:28]: same, why change from latest `Config`?

File: app/src/redux/config/__tests__/selectors.test.ts [hunk 10]
```diff
@@ -43,6 +43,20 @@ describe('shell selectors', () => {
     })
   })
 
+  describe('getHasJustUpdated', () => {
+    it('should return null if config is unknown', () => {
```
Comment (brenthagen) [inline app/src/redux/config/__tests__/selectors.test.ts]: `should return false`

File: app/src/redux/analytics/types.ts [hunk 11]
```diff
@@ -7,8 +6,9 @@ import {
 import type { CalibrationCheckComparisonsPerCalibration } from '../sessions/types'
 import type { DeckCalibrationStatus } from '../calibration/types'
 import type { Mount } from '@opentrons/components'
+import type { ConfigV0 } from '../config/types'
 
-export type AnalyticsConfig = Config['analytics']
+export type AnalyticsConfig = ConfigV0['analytics']
```
Comment (brenthagen) [inline app/src/redux/analytics/types.ts:11]: same, why not latest `Config`

**Review by brenthagen — COMMENTED** [review]
just a couple initial comments - i'm having some issues with the discovery client on this branch for probably tangential reasons. will try to review fully if i can sort those out

**Comment by mjhuff** [comment]
@brenthagen  Just throwing our discussion yesterday here to keep everyone in the loop - TypeScript is having issues at a certain depth size with determining specific config properties. By specifying the latest Config version type, the error is no longer an issue. It does solve the immediate problem, however it means a test will fail (and require updating) if a specific config category is changed. 

I did some looking into existing patterns addressing extending type implementations as we have been doing, and didn't turn up much other than "don't do it that way." One workaround is to manually type out all the properties, but I'd argue this makes things much less readable. I've got a couple other ideas I'll test out in a bit as well. 

TLDR; the current solution isn't great, but it works.

**Review by shlokamin — APPROVED** [review]
code lgtm, great work Jamey! this is some of the oldest code in the app, thanks for being down to dive in head first. 

i unfortunately won't be able to thoroughly test this on a robot until tomorrow. if someone else can/already has though im cool with merging. just wanna make sure we run this to ground so we don't interfere with internal + external beta testing
